src/mailadm/bot.py

```diff
@@ -0,0 +1,121 @@
+                    "/add-user addr password token\n"
+                    "/list-tokens")
+            message.chat.send_text(text)
+
+        elif arguments[0] == "/add-token":
+            if len(arguments) == 4:
+                arguments.append("")  # add empty prefix
+            text = add_token(self.db, name=arguments[1], expiry=arguments[2], maxuse=arguments[3],
+                             prefix=arguments[4], token=None).get("message")
+            message.chat.send_text(text)
+
+        elif arguments[0] == "/add-user":
+            arguments = message.text.split(" ")
+            result = add_user(self.db, addr=arguments[1], password=arguments[2], token=arguments[3])
+            if result.get("status") == "success":
+                user = result.get("message")
+                text = "successfully created %s with password %s" % (user.addr, user.clear_pw)
+            else:
+                text = result.get("message")
+            message.chat.send_text(text)
+
+        elif arguments[0] == "/list-tokens":
+            message.chat.send_text(list_tokens(self.db))
+
+    def check_privileges(self, command: deltachat.Message):
+        """
+        Checks whether the incoming message was in the admin group.
+        """
+        if command.chat.is_group() and self.admingrpid == str(command.chat.id):
+            if command.chat.is_protected() \
+                    and command.is_encrypted() \
+                    and int(command.chat.num_contacts()) >= 2:
+                if command.get_sender_contact() in command.chat.get_contacts():
+                    return True
+                else:
+                    print("%s is not allowed to give commands to mailadm." %
+                          (command.get_sender_contact(),))
+            else:
+                print("admin chat is broken. Try `mailadm setup-bot`. Group ID:", self.admingrpid)
+                raise ValueError
+        else:
+            return False
+
+
+def get_admbot_db_path():
+    db_path = os.environ.get("ADMBOT_DB", "/mailadm/docker-data/admbot.db")
+    try:
+        sqlite3.connect(db_path)
+    except sqlite3.OperationalError:
+        raise RuntimeError("admbot.db not found: ADMBOT_DB not set")
+    return db_path
+
+
+def main(mailadm_db):
+    ac = deltachat.Account(get_admbot_db_path())
+    if not ac.is_configured():
+        print("if you want to talk to mailadm with Delta Chat, please run: mailadm setup-bot",
+              file=sys.stderr)
+    conn = mailadm_db.read_connection(closing=False)
+    while "admingrpid" not in [item[0] for item in conn.get_config_items()]:
+        time.sleep(1)
+    else:
+        conn.close()
+        ac = deltachat.Account(get_admbot_db_path())
+        ac.run_account(account_plugins=[AdmBot(mailadm_db, ac)], show_ffi=True)
+    ac.wait_shutdown()
+    print("shutting down bot.", file=sys.stderr)
+
+
+if __name__ == "__main__":
+    mailadm_db = DB(get_db_path())
```

Comment:
for symmetry it'd be better to also call `get_admbot_db_path` here and pass it into `main`.  It's usually best if as few functions as possible (and usually top-level ones) fish information from "env vars" or other global settings, instgead of doing it from deeper-in functions.  You may want to watch https://www.destroyallsoftware.com/screencasts/catalog/functional-core-imperative-shell for more elaborate context.

Reply:
ah yes, I've seen the first one :) fixed it